File: algorithms/remove_duplicates_from_list2.cpp
/*
 * =====================================================================================
 *
 *       Filename:  remove_duplicates_from_list2.cpp
 *
 *    Description:  Remove Duplicates from Sorted List II
 *
 *        Version:  1.0
 *        Created:  02/22/19 04:43:02
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Zhu Xianfeng (), xianfeng.zhu@gmail.com
 *   Organization:
 *
 * =====================================================================================
 */
#include <cstdio>
#include <cstdlib>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// Two pointers
class Solution {
 public:
  ListNode* deleteDuplicates(ListNode* head) {
    ListNode dummy(0, head);
    ListNode* p = &dummy;
    ListNode* left = p->next;
    ListNode* right = nullptr;
    bool duplicate = false;
    while (left != nullptr) {
      right = left->next;
      if (right != nullptr && left->val == right->val) {
        left->next = right->next;
        duplicate = true;
      } else if (duplicate) {
        p->next = left->next;
        left = left->next;
        duplicate = false;
      } else {
        p = p->next;
        left = left->next;
      }
    }
    return dummy.next;
  }
};

int main(int argc, char* argv[]) {
  ListNode* root = NULL;
  root = Solution().deleteDuplicates(root);

  return 0;
}

File: algorithms/remove_duplicates_from_list2.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# 82. Remove Duplicates from Sorted List II
# https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        curr = dummy = ListNode()
        left, right = head, head.next
        while left:
            while right and right.val == left.val:
                right = right.next
            if right == left.next:
                curr.next = left
                curr = curr.next
            if not right:
                curr.next = left = None
            else:
                left, right = right, right.next
        return dummy.next
